Android call graph construction

Diarilex opened the pull request:
Adds the AndroidCallGraphKey, that can be used to construct call graphs for Android apps. 
The app must be available as a .jar file and the AndroidManifest.xml is required.

roterEmil reviewed (approved):
works for me

Diarilex committed:
add android

Diarilex committed:
add android

Diarilex committed:
version update

Diarilex committed:
Delete AndroidSchedulerTest.scala

errt commented on OPAL/br/src/main/resources/reference.conf:
```diff
@@ -59,7 +59,8 @@ org.opalj {
           #analysis = "org.opalj.br.analyses.cg.ApplicationEntryPointsFinder"
-          analysis = "org.opalj.br.analyses.cg.ApplicationWithoutJREEntryPointsFinder"
+          #analysis = "org.opalj.br.analyses.cg.ApplicationWithoutJREEntryPointsFinder"
+          analysis = "org.opalj.br.analyses.cg.AndroidEntryPointsFinder"
```
Please do not change the default configuration here

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/AndroidCallGraphKey.scala:
```diff
@@ -0,0 +1,55 @@
+package tac
+package cg
+
+import org.opalj.br.analyses.{ProjectInformationKey, SomeProject}
```
Do not use multi-imports

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/AndroidCallGraphKey.scala:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @author Tom Nikisch
+ */
+class AndroidCallGraphKey(cgk: ProjectInformationKey[CallGraph, Nothing], manifestPath: String) extends CallGraphKey {
```
This class seems to be a bad idea in general. As far as I can tell, all that it does is adding the AndroidICC call graph module to the set of executed call graph modules for either the CHA or RTA call graph. This should not be necessary, as all call graph keys can just pick up such an additional module from the "org.opalj.tac.cg.CallGraphKey.modules" anyway, so no new CallGraphKey implementation is required. Also, this design allows neither for changes in the CHA or RTA call graph keys nor can it be combined with any other call graph key. In fact, it would even lead to NullPointerExceptions if one reasonably tried to just supply a different CallGraphKey.

errt commented on OPAL/tac/src/main/resources/callbackList.txt:
```diff
@@ -0,0 +1,181 @@
+android/accounts/OnAccountsUpdateListener
```
Is this file used at all? Because I can't find it referenced anywhere. If it is necessary, please ensure that it has a proper, meaningful name that can be understood in the context of OPAL in general (i.e., currently one could only make sense of this if they already knew that it was about call graphs for Android). Finally, please ensure that such files are placed in a proper location (resources of the OPAL subproject that uses the file is fine, but a meaningful directory structure below that should be used).

errt replied:
I found the place the file is used. Still, please name it appropriately and create a meaningful directory structure for it.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -347,4 +347,46 @@ object AllEntryPointsFinder extends EntryPointFinder {
+object AndroidEntryPointsFinder extends EntryPointFinder {
+
+    val activityEPS: List[String] = List("onCreate", "onRestart", "onStart", "onResume",
+        "onStop", "onDestroy", "onActivityResult")
```
Avoid line breaks just anywhere in such an argument list. For proper formatting, start the arguments on a new line and also put the closing parenthesis on a line of its own.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -347,4 +347,46 @@ object AllEntryPointsFinder extends EntryPointFinder {
+ */
+object AndroidEntryPointsFinder extends EntryPointFinder {
+
+    val activityEPS: List[String] = List("onCreate", "onRestart", "onStart", "onResume",
```
Avoid the abbreviation EPS for entry points, as in the context of OPAL, this is used extensively to mean Entity/Property-State.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -347,4 +347,46 @@ object AllEntryPointsFinder extends EntryPointFinder {
+        "onStatusChanged")
+    val onNmeaMessageListenerEPS: List[String] = List("onNmeaMessage")
+    val broadcastReceiverEPS: List[String] = List("onReceive")
+    val defaultEPS = Map("android/app/Activity" -> activityEPS, "android/app/Service" -> serviceEPS,
```
It seems it would be sensible to have this in a configuration file instead of hardcoded. You can easily represent such a Map[String, List[String]] in the config format and load that where you need it.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -347,4 +347,46 @@ object AllEntryPointsFinder extends EntryPointFinder {
+    def findEPS(ot: ObjectType, possibleEPS: List[String], project: SomeProject): ArrayBuffer[Method] = {
+        val eps = ArrayBuffer.empty[Method]
+        val classHierarchy = project.classHierarchy
+        classHierarchy.foreachSubclass(ot, project) { sc ⇒
```
Probably correct, but just to make sure: This is not reflexive, i.e., this will not be executed for the given ObjectType, but ONLY for its subclasses. It is also not executed for any interface types.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -347,4 +347,46 @@ object AllEntryPointsFinder extends EntryPointFinder {
+        val eps = ArrayBuffer.empty[Method]
+        val classHierarchy = project.classHierarchy
+        classHierarchy.foreachSubclass(ot, project) { sc ⇒
+            for (pep ← possibleEPS; m ← sc.findMethod(pep) if m.body.isDefined && !eps.contains(m)) {
```
Please use the multi-line for syntax with braces {} instead of the single-line with parentheses ().

errt replied:
While probably not too important here, it is quite imperformant to do contains on an ArrayBuffer. Since what you're obviously after is the semantics of a Set (i.e., you only add elements once anyway), why don't you use a Set or Set-like data structure in the first place?

errt replied:
Just to be sure: Are you aware that findMethod with only a method name will return all methods of the given name? That is probably fine if the methods that you are looking for can come with different signatures for the same name and still be entry points to the call graph and if it is unlikely that someone will implement a method with one of these names that is not an entry point. On the other hand, if there is only one (or in few cases maybe two or three) possible signature anyway and if it is conceivable that someone would have non-entry point methods of those names, I would suggest to also include these signatures in the query here.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+    //explicit
+    final val setClass = "setClass"
+    final val setClassName = "setClassName"
+    final val setComponent = "setComponent"
```
I don't see a proper reason for all of these trivial string definitions.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+    def performAnalysis(
+                         manifest: Elem
+    ): ProperPropertyComputationResult = {
+        parseManifest(manifest)
```
I would strongly suggest a more functional programming style here instead of imperatively invoking several methods that manipulate some state that probably isn't required elsewhere.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/AndroidCallGraphKey.scala:
```diff
@@ -0,0 +1,55 @@
+    override protected def callGraphSchedulers(
+        project: SomeProject
+    ): Traversable[FPCFAnalysisScheduler] = {
+        val AndroidICC = eagerAndroidICCAnalysisScheduler
```
Use proper camelCase (i.e., start with a small letter) for local variables. Only classes should start with a capital letter.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+import scala.io.Source
+import scala.xml.{Elem, XML}
+
+sealed trait AndroidICCPropertyMetaInformation extends PropertyMetaInformation {
```
What is this property? It seems to be unused (and useless anyway, since you're just computing call graph data).

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+    }
+    override def requiredProjectInformation: ProjectInformationKeys = Seq(ComputeTACAIKey)
+
+    override def uses: Set[PropertyBounds] = Set(PropertyBounds.lb(Callers), PropertyBounds.lb(Callees))
```
Use ProperyBounds.lbs(...) to create multiple at once.  However, this should almost certainly be ubs anyway (a lower bound is an overapproximation as an interim result, however, our call graphs usually produce underapproximations. I.e., they add edges as the computation progresses, not remove any).

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+ * In order for the analysis to work the path to the projects AndroidManifest must be set via setManifest
+ */
+object eagerAndroidICCAnalysisScheduler extends BasicFPCFEagerAnalysisScheduler {
+    var manifestPath: String = _
```
A mutable, uninitialized field like this in an object is a disaster waiting to happen. Please consider solving your problem differently.

errt replied:
I think the proper way to do this is to create a ProjectInformationKey that does the manifest parsing. That key can get the manifest via its InitializationData. If none has been given, it should also work (for cases in which no manifest is available).

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+    }
+}
+
+class implicitIntent(var caller: Method) {
```
Use proper CamelCase (i.e., start with a capital letter) for classes.

errt replied:
Should clearly be documented.

errt replied:
Should maybe be an inner class in the analysis itself as I don't think it is used outside of the analysis.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+    var iData: String = ""
+    var componentTypes: ListBuffer[String] = ListBuffer.empty[String]
+    var calledMethods: ListBuffer[String] = ListBuffer.empty[String]
+    var pc: Int = 0
```
All of this looks very strange. Should this maybe be a case class? Why are there setter methods for some fields but not for others?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+    ): Unit = {
+        val statements = tacCode.stmts
+        useSites.foreach { ln ⇒
+            val stm = statements(ln)
```
Please use "stmt" as the common abbreviation used for "statement" in OPAL

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+                          useSites: List[Int]
+    ): Unit = {
+        val statements = tacCode.stmts
+        useSites.foreach { ln ⇒
```
What is "ln" supposed to mean? As you're iterating over useSites, I'd usually expect the variable here to be called "use", though somethig more specific may also be used to make things clearer.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+                }
+            }
+            if (stm.isAssignment) {
+                if (stm.asAssignment.expr.asVirtualFunctionCall.name == registerReceiver) {
```
Looks like an exception waiting to happen. What if the expr isn't a VirtualFunctionCall?

errt replied:
Should obviously extract stm.asAssignment.expr.asVirtualFunctionCall into a local variable to avoid the repeated use.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+    }
+}
+
+class intentFilter() {
```
See comments on implicitIntent below.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+
+}
+
+class explicitIntent(val caller: Method, val callee: ClassFile) {
```
See comments on implicitIntent below.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+        }
+    }
+
+    def overApproximateEdges(
```
Seems to be used only once, so better not have a separate method for such a small functionality.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+                    matchFound(intent, filter)
+                }
+            } else {
+                if (data.nonEmpty) {
```
This ends up extremely deeply nested. Can you maybe refactor the code to avoid the deep nesting levels? E.g., it should be possible to directly use else if here instead of an else that is followed by a nested if.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+        }
+    }
+
+    def testData(
```
Please document methods like these. Also, a more meaningful naming of methods seems to be possible. It is totally unclear to me what, e.g., testData, would mean or do.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+                          superTypes: UIDSet[ObjectType],
+                          filter: intentFilter
+    ): Unit = {
+        if (superTypes.contains(ObjectType(androidActivity))) {
```
What if superTypes contains more than one of these?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+                val intent = new implicitIntent(m)
+                evaluateUseSites(tacCode, useSites, intent)
+            case 1 ⇒ //Intent(String action)
+                if (p.head.asVar.value.verificationTypeInfo == ObjectVariableInfo(ObjectType(stringString))) {
```
There are constants available for ObjectTypes like java/lang/String: ObjectType.String. Please always use these as it is way more performant than looking them up by their name each time. For types that don't have constants in OPAL but are needed by you several times, consider looking up the ObjectType only once and store it as a constant in your analysis.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+                             tacCode: AITACode[TACMethodParameter, ValueInformation]
+                           ): Option[(ObjectType, String)] = {
+        functionCall.params.foreach { p ⇒
+            if (p.asVar.value.toString.startsWith(intentStart.head) || p.asVar.value.toString.startsWith(intentStart(1))) {
```
You should probably extract p.asVar.value.toString to a local variable to shorten this.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+
+                    val call = defSiteStm.asAssignment.expr.asFunctionCall
+                    if (!intentMethods.contains(call.name)) {
+                        return Some[(ObjectType, String)](call.declaringClass.asObjectType -> call.name)
```
Is the explicit type here necessary? I think it should be able to infer it.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+        }
+    }
+
+    //Checks if the intent is reachable within the processed method
```
Please always use proper ScalaDoc comments for your documentation

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+    }
+
+    /**
+     * searches the parameters of functionCall to find the origin of the intent and return the class and method
```
Please format your ScalaDoc comments: Descriptions should start with a captial letter (as any proper sentence) and end with a full stop. Remove blank lines that are unnecessary.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+                        }
+                    }
+                }
+                return
```
Do not use void returns, they should not be necessary here.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+                val stm = tacCode.stmts(defIndex)
+                if (stm.isAssignment) {
+                    if (stm.asAssignment.expr.isStaticFunctionCall || stm.asAssignment.expr.isVirtualFunctionCall) {
+                        return findOrigin(stm.asAssignment.expr.asFunctionCall.params, tacCode, originalUseSites)
```
If you really have to use an early return (which should be avoided where possible), please add a semicolon at its end. This makes it stand out visually.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1413 @@
+        if(fullLifecycle) {
+            val activityCF = project.classFile(ObjectType(androidActivity)).orNull
+            if (activityCF != null) {
+                val AonCreate = activityCF.findMethod(onCreate).head
```
Proper camelCase for local variables please.

errt reviewed (changes requested):
Partial review only, a more thorough review might be in order once the comments have been addressed. Please use `sbt scalariformFormat` to format your code, it can not be merged without proper formatting.

Diarilex committed:
default configuratio, multi imports

Diarilex committed:
delete AndroidCallGraphKey.scala

Diarilex committed:
Merge remote-tracking branch 'origin/develop' into develop

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1430 @@
+    }
+
+    def internalFinder(
+                        nme:       String,
```
This is "name", right? If so, please write it out. No point in making things less clear for one saved letter.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1430 @@
+     *
+     */
+    def searchIntentsAndCallbacks(): Unit = {
+        val tacKey = project.get(ComputeTACAIKey)
```
Don't use ComputeTACAIKey, please get the TACAI through the PropertyStore (see other call graph analyses for reference)

Diarilex replied:
I tried:
val tacCode = propertyStore.get(m, TACAI.key).get.ub.tac.get
It mostly does not provide any TACAI. How can i ensure that the required information is available?

errt replied:
You have to match the result of propertyStore(m, TACAI.key) (don't use get, just use the apply) to check whether it is available. If it is not yet available, you have to return an InterimResult and receive the TACAI later in a continuation function. Take a look at, e.g., LoadedClassesAnalysis, lines 100ff for an example.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1430 @@
+     */
+    def searchIntentsAndCallbacks(): Unit = {
+        val tacKey = project.get(ComputeTACAIKey)
+        val source = Source.fromFile("OPAL\\tac\\src\\main\\resources\\callbackList.txt")
```
Use Source.fromInputStream together with this.getClass.getResourceAsStream

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1430 @@
+            val tacCode = tacKey(m)
+
+            body.collect {
+                case in: Instruction if in.isInstanceOf[INVOKESPECIAL] ⇒ in
```
I think this should work:
```
body.foreach{ case in: INVOKESPECIAL =>
[...]
}
```
Alternatively, this could be a solution:
```
collectWithIndex { case PCAndInstruction(pc, _: INVOKESPECIAL) =>
[...]
}
```

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1430 @@
+        val p = c.params
+
+        p.size match {
+            case 0 ⇒ //Intent()
```
Can't you match the actual signature of the call?

Diarilex replied:
Yes this may also provide more accurate results. Can I use the method descriptor for this?

errt replied:
Yes, c.descriptor gives you the signature (a.k.a. descriptor) of the method called.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1430 @@
+     *
+     */
+    def performAnalysis(
+                         manifest: Elem
```
I don't think the manifest should be the entity here. Maybe the project is the better idea here? Call graph modules are usually built to be executed once for each reachable method though. Would that be possible here?

Diarilex replied:
The analysis makes additional methods reachable after it is done (it collects all information about intents and filters and processes them to find the emerging calls), which would then "restart" the analysis and therefore add information that needs to be processed. I suppose that this would be less efficient, because the analysis would need to be executed multiple times. Also the analysis always has to compute all found information, not only the newly found ones.
I will change the entity to the project, once i am done with the ProjectInformationKey for the Manifest parsing.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1430 @@
+
+    override def derivesEagerly: Set[PropertyBounds] = Set.empty
+
+    override def derivesCollaboratively: Set[PropertyBounds] = Set(PropertyBounds.lb(Callers), PropertyBounds.lb(Callees))
```
Here it should also be ubs!

Diarilex committed:
Various changes addressing the comments.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -357,36 +357,71 @@ object AllEntryPointsFinder extends EntryPointFinder {
+    )
+    val contentProviderEntryPoints: List[(String, MethodDescriptor)] = List(
+        ("onCreate", MethodDescriptor.NoArgsAndReturnVoid),
+        ("query", null),
```
What are these null descriptors? Are these methods where multiple descriptors apply? If it is only two or three, maybe still just list them all. If you need to match all descriptors, I'd prefer to change the type here to Option[MethodDescriptor] and have None instead of null for more idiomatic Scala.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -5,10 +5,7 @@ package fpcf
-import org.opalj.br.Method
-import org.opalj.br.ObjectType
-import org.opalj.br.ObjectVariableInfo
+import org.opalj.br.{ClassFile, Method, ObjectType, ObjectVariableInfo, PCAndInstruction}
```
Please avoid these multi-imports. You should configure your IDE to not create them.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -599,132 +578,131 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
     ): Unit = {
         val p = c.params
-
+        println(c.descriptor)
```
Don't forget to remove this.

errt reviewed (commented):
Thanks for renaming the CallbackList textfile, but please change the directory to actually be subdirectories instead of one directory with dots in its name.

Diarilex committed:
Added ParseManifestKey and solved more issues.

Diarilex committed:
get TACAI from propertyStore

Diarilex committed:
minor refactor

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1346,17 +1202,194 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+    }
+}
+
+object ManifestParsingKey extends ProjectInformationKey[(Map[String, ListBuffer[ClassFile]], ListBuffer[IntentFilter]), Elem] {
```
Please move this to a file of its own. Probably outside of the call graph packages, as this could be useable for other purposes as well. Maybe an android package would be a good idea?

Diarilex replied:
Where should this be located? org.opalj.tac.fpcf.analyses.cg.android or more like org.opalj.tac.android?

errt replied:
What exactly does it do? Is it only useful for call graphs, then *.cg.android is fine. Otherwise, a more top level package seems reasonable. Also, this class doesn't use any functionality from the TAC, right? In that case, it should probably be somewhere in org.opalj.br (the bytecode representation subproject).

Diarilex replied:
It uses functionalities from the AndroidICCAnalysis (IntentFilter). The analysis collects intent filter and Android components from the Manifest, which might not be useful for other analyses apart from call graphs.

errt replied:
But intent filters are also not necessarily only useful for the call graph directly, are they? They do contain information about the types of intents and such that might be useful elsewhere in the future, e.g., for just checking what kinds of intents are used without a full call graph analysis? If that is the case, the intent filters should also be moved out of the call graph hierarchy.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1346,17 +1202,194 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+
+    override def compute(project: SomeProject): (Map[String, ListBuffer[ClassFile]], ListBuffer[IntentFilter]) = {
+        val manifest = project.getProjectInformationKeyInitializationData(ManifestParsingKey)
+        if (manifest.isDefined) {
```
Use manifest.map to create an `Option`al return value instead of using null.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1346,17 +1202,194 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+    }
+}
+
+object ManifestParsingKey extends ProjectInformationKey[(Map[String, ListBuffer[ClassFile]], ListBuffer[IntentFilter]), Elem] {
```
I think `AndroidManifestKey` or something similar might be a better name.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -357,37 +357,37 @@ object AllEntryPointsFinder extends EntryPointFinder {
+    val broadcastReceiverEntryPoints: List[(String, Option[MethodDescriptor])] = List(
+        ("onReceive", Some(MethodDescriptor.apply(FieldTypes(FieldType("Landroid/content/Context;"), FieldType("Landroid/content/Intent;")), VoidType)))
     )
     val defaultEntryPoints = Map(
```
Still wondering whether it would be better to encode all of this in the reference.conf, then parse it here.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -374,140 +315,144 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
     def searchIntentsAndCallbacks(): Unit = {
-
-        val source = Source.fromInputStream(this.getClass.getResourceAsStream("/org.opalj.tac.fpcf.analyses.cg/AndroidCallbackList.txt"))
+        val source = Source.fromInputStream(this.getClass.getResourceAsStream("/org/opalj/tac/fpcf/analyses/cg/AndroidCallbackList.txt"))
```
I believe it should be possible to not use a full path here. Could you please try?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -374,140 +315,144 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+                                if (p.isObjectType) {
+                                    val ot = p.toString.replaceAll(otRegex, "")
+                                    //only interfaces defined in 'callbacklist' are handled as callback interfaces and dynamically searching for callback interfaces
+                                    if (callbackList.contains(ot) ||
```
contains on a (large) list is slow, it is probably better to turn the list into a set right away.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -374,140 +315,144 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+                            var i = 0
+                            methodCall.asVirtualMethodCall.descriptor.parameterTypes.foreach { p ⇒
+                                if (p.isObjectType) {
+                                    val ot = p.toString.replaceAll(otRegex, "")
```
I don't understand what you're trying to do with this regex. Turning a proper object into its string representation, manipulating that and then using that to lookup something seems like a bad idea in general. Can't you achieve the same thing directly using the ObjectType objects?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -374,140 +315,144 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+                                    //only interfaces defined in 'callbacklist' are handled as callback interfaces and dynamically searching for callback interfaces
+                                    if (callbackList.contains(ot) ||
+                                        (ot.startsWith("android") && ot.endsWith("Listener") && ot.substring(ot.lastIndexOf("$") + 1).startsWith("On"))) {
+                                        val par = methodCall.asVirtualMethodCall.params(i).asVar.value.verificationTypeInfo
```
What are you trying to achieve here? The use of verificationTypeInfo seems really odd. Maybe what you need is something like the upper type bound (or all types) that you can get from value.asReferenceValue?

Diarilex replied:
I am using the verificationTypeInfo to get the ObjectType of the parameter(i) of the methodCall, because i need to get the classFile. Is there a more elegant way to access the classFile of this parameter?

errt replied:
As I said, value.asReferenceValue.upperTypeBound might be the way to do it.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -374,140 +315,144 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+                        val methodCall = tacCode.stmts(index)
+                        if (methodCall.isVirtualMethodCall) {
+                            var i = 0
+                            methodCall.asVirtualMethodCall.descriptor.parameterTypes.foreach { p ⇒
```
I think you can use foreachWithIndex here to get rid of `i`.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -374,140 +315,144 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+            if (callbackMethods.nonEmpty) {
+                generateCallBackEdges(m, callbackMethods)
+            }
+        } catch {
```
This empty catch seems like a really bad idea. What exceptions do you expect here? Is it really ok to not handle them at all? Should there at least be a warning of some kind?

Diarilex replied:
It was meant to catch methods, where no tac representation  was available. It is no longer necessary, as the tac representation is provided by the property store.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -658,58 +603,61 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
-            case 0 ⇒ //Intent()
+
+        c.descriptor match {
+            case MethodDescriptor(types, _) if types.isEmpty ⇒ //Intent()
```
Since you only distinguish by the types anyway, shouldn't you do `c.descriptor.parameterTypes match`?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -658,58 +603,61 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
                 evaluateUseSites(tacCode, useSites, intent)
-            case 1 ⇒ //Intent(String action)
+
+            case MethodDescriptor(types, _) if types == FieldTypes(FieldType("Ljava/lang/String;")) ⇒
```
Do not create types on the fly. Use, e.g., ObjectType.String for predefined types. Define other types (like the Intent) once on the class level. Also, do not use FieldType("L[...];"), but ObjectType("[...]") directly.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -658,58 +603,61 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+            case MethodDescriptor(types, _) if types == FieldTypes(FieldType("Landroid/content/Context;"), FieldType("Ljava/lang/Class;")) ⇒
+                //Intent(Context context, Class cls)
+                val clsFile = project.classFile(ObjectType(p.last.asVar.value.toString.
+                    replaceAll(valueRegex, "").replaceAll("\\.", "/")))
```
Again with all of those regexes. Can't you use the object oriented representation directly instead of manipulating its string representations?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -658,58 +603,61 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
-                        }
+                if (clsFile.isDefined) {
+                    val intent = new ExplicitIntent(m, clsFile.get)
+                    intent.findStartedMethod(tacCode, useSites)
```
This seems like a bad pattern (calling a mutating method on the Intent object. Is that startedMethod always needed anyway? Then it should probably be initialized by the constructor of the Intent object.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -658,58 +603,61 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+                if (clsFile.isDefined) {
+                    val intent = new ExplicitIntent(m, clsFile.get)
+                    intent.findStartedMethod(tacCode, useSites)
+                    if (intent.calledMethod.nonEmpty) {
```
There seems to be a naming inconsistencey between `findStartedMethod` and `calledMethod`, as it seems both refer to the same method?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1346,17 +1202,194 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+        addDataAuthority, addDataPath, addDataScheme, addDataSSP, addDataType
+    )
+
+    var receiver: ClassFile = _
```
Try to avoid uninitilized variables, there is almost never a need for them.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1346,17 +1202,194 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+
+    var receiver: ClassFile = _
+    var componentType: String = _
+    var registeredReceivers: ListBuffer[ObjectType] = ListBuffer.empty[ObjectType]
```
It seems weird to have mutable collections in a mutable variable. Usually you'd have either a mutable collection in an immutable val or the other way round.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1346,17 +1202,194 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+        )
+        val intentFilters = ListBuffer.empty[IntentFilter]
+        val androidURI = "http://schemas.android.com/apk/res/android"
+        val pack = manifest.attribute("package").get.toString().replaceAll("\\.", "/")
```
Avoid using short names, e.g., `packageName` would probably be much clearer here.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1346,17 +1202,194 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+                            intentFilter.receiver = rec.get
+                            intentFilter.actions = (filter \ "action").map(_.attribute(androidURI, "name").
+                                get.head.toString()).to[ListBuffer]
+                            intentFilter.categories = (filter \ "category").map(_.attribute(androidURI, "name").
```
You probably want to move all this initialization to be parameters of the IntentFilter class (i.e., constructor parameters).

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1346,17 +1202,194 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
-    def setManifest(manifestPath: String): Unit = {
-        this.manifestPath = manifestPath
-    }
+
```
Class name should start with a capital letter.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -319,12 +314,22 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
-            }
+            } else leftOverMethods += m
         })
+        while (leftOverMethods.size > 1) {
```
This busy waiting is an extremely bad idea. You absolutely have to handle this using a continuation function.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -435,7 +440,7 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
                                             val cf = project.classFile(par.asObjectVariableInfo.clazz.asObjectType)
                                             if (cf.isDefined) {
-                                                val methods = cf.get.methods.filter(f ⇒ f.name != init && f.name != "<clinit>")
+                                                val methods = cf.get.methods.filter(f ⇒ !f.isConstructor && f.name != "<clinit>")
```
Use f.isStaticInitializer instead of checking for "<clinit>" yourself. In fact, use f.isInitializer to check for both constructors and static initializers at once.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -494,7 +499,13 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+     * @param params Parameters of the analysed statement.
+     * @param tacCode TAC representation of the statement.
+     * @param originalUseSites UseSites of the statement.
+     * @return
```
Please document the return value as well.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -494,7 +499,13 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
 
-    //Checks if the intent is reachable within the processed method
+    /**
+     * Checks if the intent is reachable within the processed method
```
Proper method descriptions should end in a full stop.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1387,10 +1404,12 @@ object ManifestParsingKey extends ProjectInformationKey[(Map[String, ListBuffer[
 /**
  * Schedules the execution of an AndroidICCAnalysis.
  * In order for the analysis to work the path to the projects AndroidManifest must be set via setManifest.
```
This comment is really missing an imporant comma: `to work, the path`. Though I don't think it is still valid, now that the manifest is parsed by a project information key?

errt reviewed (changes requested):
Code quality is improving, but there is still a bunch of problems.

Diarilex committed:
Made AndroidICCAnalysis a two phase analysis.
Moved the AndroidManifestKey to tac/cg/android.
Additional minor fixes.

Diarilex committed:
verificationTypeInfo => upperTypeBound

Diarilex committed:
format

Diarilex committed:
Moved the Android entry points to the reference.conf

Diarilex committed:
Merge branch 'develop' into develop

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidManifestKey.scala:
```diff
@@ -0,0 +1,101 @@
+import org.opalj.br.analyses.ProjectInformationKey
+import org.opalj.br.analyses.ProjectInformationKeys
+import org.opalj.br.analyses.SomeProject
+import org.opalj.tac.fpcf.analyses.cg.IntentFilter
```
It seems that IntentFilter should be moved to the cg.android package as well.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -307,155 +323,165 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+        } else Results(partialResults)
+    }
+
+    private[this] def continuationForTAC(m: Method)(eps: SomeEPS): PropertyComputationResult = {
```
I'm confused. Aren't continuations required to return a `ProperPropertyComputationResult`, i.e., not `NoResult`?

Diarilex replied:
It did work with NoResult, but it might cause trouble I am not aware of. At this point there is no result available. Is there any way I can return a ProperPropertyComputationResult, that acts like NoResult?

errt replied:
It seems I was wrong and InterimPartialResults may actually return NoResult. So it should not be a problem.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -6,15 +6,12 @@ package analyses
 package cg
```
Move this to the cg.android package as well?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -307,155 +323,165 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
     }
 
-    private def processMethod(m: Method, callbackList: List[String]) = {
+    private def processMethod(m: Method): PropertyComputationResult = {
```
Why does this return a `PropertyComputationResult` when that is always `NoResult` and seems to be unused?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -830,57 +845,65 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+            try {
+                project.classFile(tacCode.stmts(defSite).asAssignment.expr.asClassConst.value.asObjectType)
+            } catch {
+                case _: Throwable ⇒ None
```
Looks fishy, please write your code in such way that it doesn't cause an exception in the first place.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala:
```diff
@@ -1297,117 +1317,24 @@ class IntentFilter() {
-                        }
+                    if (virtualCall.receiver.asVar.definedBy.head > -1 && statements(virtualCall.receiver.asVar.definedBy.head).asAssignment.expr.isGetField) {
+                        registeredReceivers += statements(virtualCall.receiver.asVar.definedBy.head).asAssignment.expr.asGetField.objRef.asVar.value.verificationTypeInfo.asObjectVariableInfo.clazz.asObjectType
+                        //statements(virtualCall.receiver.asVar.definedBy.head).asAssignment.expr.asGetField.objRef.asVar.value.toString.replaceAll(receiverRegex, "").replaceAll("\\.", "/"))
```
If that code is not needed anymore, please just remove it.

errt commented on OPAL/br/src/main/resources/reference.conf:
```diff
@@ -100,6 +100,42 @@ org.opalj {
           ]
         }
+
+        DefaultEntryPoints{
```
Please use a name that makes clear the connection to Android. In general, it should be the same name (and package structure) as the analysis/ProjectInformationKey that uses it.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -357,55 +359,37 @@ object AllEntryPointsFinder extends EntryPointFinder {
     override def collectEntryPoints(project: SomeProject): Traversable[Method] = {
         val entryPoints = ArrayBuffer.empty[Method]
+
+        val defaultEntryPoints = Map(
```
Better, but I'm not sure an immutable map with mutable ListBuffers is a good idea. You should be able to just iterate over the keys in the config entry (and if that is not possible, start with a list of the keys here and create the map on the fly when parsing the config to create a fully immutable map.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -357,55 +359,37 @@ object AllEntryPointsFinder extends EntryPointFinder {
 
-    def findEntryPoints(ot: ObjectType, possibleEntryPoints: List[(String, Option[MethodDescriptor])], project: SomeProject): Set[Method] = {
+    def parseMethodDescriptor(descriptor: String): Option[MethodDescriptor] = {
+        if (descriptor != "None") {
```
It might be better to, instead of explicitly putting the String "None" into the config, just leave it out. I think you should be able to get an optional value for the descriptor than that you can just `descriptor.map(MethodDescriptor(_))` then.

errt reviewed (changes requested):
We're getting closer. If the open comments are addressed, I think this could just benefit from some more documentation (there are some long stretches of rather complex code where it's not easy to see what is done for what reasons) and then this should hopefully be ready to be merged.

Diarilex committed:
Moved AndroidICCAnalysis to the android.cg package.
Moved to AndroidCallbackList.txt to the resources of the android package.
Changed the parsing of the EntryPointsMap.

Diarilex committed:
Format

Diarilex committed:
Merge branch 'develop' into develop

Diarilex committed:
Fixed regex, added comments.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -361,20 +363,16 @@ object AndroidEntryPointsFinder extends EntryPointFinder {
 
     override def collectEntryPoints(project: SomeProject): Traversable[Method] = {
```
Please refactor this method again. I see no reason to create multiple mutable datastructures to hold intermediate values. It should be possible to just create the entryPoints as an immutable list directly from the config values.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -361,20 +363,16 @@ object AndroidEntryPointsFinder extends EntryPointFinder {
-        defaultEntryPoints.keys.foreach { d ⇒
+        val defEntry = project.config.getConfig("org.opalj.br.analyses.cg.AndroidEntryPointsFinder")
+        defEntry.entrySet().forEach{ e =>
+            val d = e.getKey.replaceAll("\"", "")
```
This replaceAll seems strange. Shouldn't the keys come without quotation marks already?

Diarilex replied:
Unfortunately they do come with quotation marks and I have not found another way around it yet.

errt replied:
You're right, and https://github.com/lightbend/config/issues/524 explains this. As far as I can tell, the correct way (see https://lightbend.github.io/config/latest/api/com/typesafe/config/Config.html section Iteration) thus should be `getConfig([...]).root()` which then is a proper map from unquoted keys to values.

errt commented on OPAL/br/src/main/scala/org/opalj/br/analyses/cg/EntryPointFinder.scala:
```diff
@@ -361,20 +363,16 @@ object AndroidEntryPointsFinder extends EntryPointFinder {
 
```
We should probably extract this class to a class file of its own and put that into the android package so that the android specific code is in a single place. Remember to change the config file so that the path for the configuration for this class stays in sync with the package the class is in.

Diarilex replied:
EntryPointFinder is a sealed trait. The AndroidEntryPointsFinder can not be moved to tac.cg.android and extend EntryPointFinder.

errt replied:
Hm, right. However, it doesn't seem that it is necessary to have that sealed, so we can probably remove that modifier.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -1,10 +1,7 @@
-package fpcf
-package analyses
-package cg
+package org.opalj.tac.cg.android
```
Fix package definition. It should have `package org.opalj` as the first line and then every subpackage on a line of its own. Please check your other files as well.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -708,7 +732,7 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
 
     /**
-     * Searches the value of a data parameter.
+     * Searches the value of a data parameter. If nothing is found returns 'undef'.
```
Returning a String with a special meaning seems like a bad idea. Should this instead return an `Option[String]`?

errt replied:
`checkData` seems not to match the actual functionality of the method. Please pick a name that makes it easy to understand what the meaning of the result value is.

Diarilex replied:
I am using the 'undef' String to mark values that could not be found, to overapproximate the intent matching at these points. If necessary I think it is possible to change  all 'undef' values to None.

errt replied:
Does this "undef" appear in the final output or do you use this just internally? If you use it only internally, None should be better to signify in the types already that the might not be a value.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+                        val defSite2 = expr.asStaticFunctionCall.params.head.asVar.definedBy.head
+                        if (defSite2 > -1) {
+                            val str = tacCode.stmts(defSite2)
+                            if (str != null &&
```
These null checks here and two lines down seem strange. Are you sure that these could ever be null? If not, please do not place spurious null checks.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+                }
+            }
+        }
+        //toDo: Find Data, check defSites until Data string is found
```
What about this TODO? Should this still be handled?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -669,7 +667,10 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
                 if (p.head.asVar.value.verificationTypeInfo == ObjectVariableInfo(ObjectType.String)) {
                     val intent = new ImplicitIntent(m)
-                    intent.action = p.head.asVar.value.toString.replaceAll(paramRegex, "")
+                    val defSite = p.head.asVar.definedBy.head
```
Are you absolutely sure that checking only the first definition site (`head`) is sufficient here? Usually, one has to check all definition sites of a variable (including some handling of the case that it is unknown because it is not inside the method, i.e., it is <0).

errt replied:
Can you comment on this? Code like this is in several places in this analysis and it is seldomly correct to check only a single definition site (unless you know you're looking for a very specific pattern that is highly unlikely to ever change in bytecode).

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -731,6 +738,15 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+        if (expr.isStringConst) {
+            Some(expr.asStringConst.value)
+        } else if (expr.isGetStatic) {
+            Some(expr.asGetStatic.name)
```
This seems very strange. The name of a static field and the value of a String constant are usually two very different things. Are you sure that this is correct here (and that what you would need is not, e.g., the actual value stored in the field instead of the field's name)?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -1338,6 +1373,18 @@ class IntentFilter(var receiver: ClassFile, var componentType: String) {
         clone
     }
 
+    def findString(tacCode: TACode[TACMethodParameter, DUVar[ValueInformation]], defSite: Int): Option[String] = {
```
This method is duplicated, please refactor to a single instance accessible from all places where you need it.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidManifestKey.scala:
```diff
@@ -0,0 +1,100 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj.tac.cg.android
```
Fix package declaration

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+    )
+    final val explicitMethods = List(setClass, setClassName, setComponent)
+
+    val intentFilters: ListBuffer[IntentFilter] = ListBuffer.empty[IntentFilter]
```
All of these mutable `ListBuffer` and `Map`s seem strange. Please consider refactoring to immutable data structures that are initialized a single time. Either in the constructor of this class, or if that is not possible, maybe encapsulate these data structures into an immutable state class as an inner class here. Avoid mutable data structures as far as possible.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+        intentMatching()
+
+        val decMeths = project.get(DeclaredMethodsKey)
+        val contexts = project.get(SimpleContextsKey)
```
Do not use the `SimpleContextsKey` directly. Instead you should probably use `typeProvider.newContext`.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+
+        val decMeths = project.get(DeclaredMethodsKey)
+        val contexts = project.get(SimpleContextsKey)
+        val resList = ListBuffer.empty[PartialResult[_, _ >: Null <: Property]]
```
Do not use a (mutable) data structure here, just declare this as `TraversableOnce` to ensure that any iterators (in `partialResults` are not evaluated but kept lazy.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+        var sLifecycleList = List.empty[Method]
+        //find default super methods
+        if (fullLifecycle) {
+            val activityCF = project.classFile(androidActivityOT).orNull
```
Do not use `orNull` and then check for null, but directly use the `Option` and use `isDefined` and `get`.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+        if (fullLifecycle) {
+            val activityCF = project.classFile(androidActivityOT).orNull
+            if (activityCF != null) {
+                val aOnCreate = activityCF.findMethod(onCreate).head
```
Are you sure that the first method (`head`) is the right thing to use here? What if there are multiple methods of that name? Maybe they all have to be analyzed? Or does this apply only to a specific method with a known descriptor? In that case, also use the descriptor to ensure that you only find that one method.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+                for (index ← 1 until lifecycleList.size) {
+                    edges += (lifecycleList(index - 1), 0) -> lifecycleList(index)
+                }
+                makeEdge(lifecycleList(3), lifecycleList(2))
```
It is totally unclear what happens here with all of these magic numbers.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+        }
+    }
+
+    def makeEdge(
```
Method signature should fit into one line of code.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+    }
+
+    def makeEdge(
+        org: Method,
```
I'd prefer it if you used more common terminology. These are call edges? Then `src`/`source` and `tgt`/`target` are common.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+        org: Method,
+        rec: Method
+    ): Unit = {
+        edges += ((org, 0) -> rec)
```
Please adopt a more functional style. Do not manipulate global data, but use local data and pass it to the method. Usually, it should also be immutable data structures and the result thus should be a new data structure that is returned by the method.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+     */
+    private def processMethod(m: Method): Unit = {
+        val body = m.body.get
+        val intentUseSites: mutable.Map[UByte, List[UByte]] = mutable.Map.empty[UByte, List[UByte]]
```
`UByte` is almost certainly the wrong type here. IntelliJ sometimes messes this up, so maybe that was the issue here, but use sites should probably be `Int`. Also, as everywhere, please avoid the use of mutable data structures.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+
+        //find sending methods (virtual)
+        body.collect {
+            case in: Instruction if in.isInstanceOf[INVOKEVIRTUAL] &&
```
Should be able to do `case in: INVOKEVIRTUAL` if senderClasses.contains(in.declaringClass) => in`. Also, I'm not sure why you don't move the check for the name here as well. You could probably do away with the `collect` here completely (that creates an intermediate data structure` and just do `foreach` (you can pattern match inside a foreach as well).

errt replied:
By the way, why are you working on the raw bytecode instructions in the body at all? Could you do this using the TAC?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+
+                if (methodCall.isVirtualMethodCall) {
+                    val virtualCall = methodCall.asVirtualMethodCall
+                    if (intentUseSites.isEmpty || (!intentUseSites.exists(k ⇒
```
Please try to keep the lamda here on one line. You should also be able to get rid of the `k` (just do `exists(_.[...]`) and of the parentheses around the `!intentUseSites.exists([...]` as well.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+    }
+
+    /**
+     * Checks if the intent is reachable within the processed method.
```
Consider using different terminology. "reachable" sounds like dead vs. live code, but this doesn't seem to be what is meant here.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+    def findOrigin(
+        params:           Seq[Expr[DUVar[ValueInformation]]],
+        tacCode:          TACode[TACMethodParameter, DUVar[ValueInformation]],
+        originalUseSites: mutable.Map[Int, List[Int]]
```
Avoid mutable data structures.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+                val index = tacCode.pcToIndex(in.pc)
+                val methodCall = tacCode.stmts(index)
+
+                if (methodCall.isVirtualMethodCall) {
```
What about `VirtualFunctionCall`s? Can't that happen as well? (To be clear: A method call is a call of a `void` method, a function call (inside an `Assignment` or `ExprStmt` is a call to a non-void method)

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+                            if (u > -1) {
+                                val stmt = tacCode.stmts(u)
+                                if (stmt.isNonVirtualMethodCall && stmt.asNonVirtualMethodCall.declaringClass ==
+                                    ObjectType("android/content/ComponentName")) {
```
Should probably pre-define that ObjectType (e.g., at the class scope) to avoid its recreation every time this function is called.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+     */
+    def intentMatching(): Unit = {
+        //ordering Filters in maps for matching
+        val filterMap: mutable.Map[String, mutable.Map[String, mutable.Map[ListBuffer[String], ListBuffer[IntentFilter]]]] = mutable.Map.empty
```
Please avoid all of these mutable data structures. Also, consider making the method more functional, not accessing and writing global state.

Diarilex replied:
Is it better to use 'var filterMap: immutable.Map[...]'?

errt replied:
Yes, it usually is. In many cases, you can even have an immutable data structure in a val if you restructure your code to build the data structure in a single pass. Might not easily be possible here, though.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+        if (uri.substring(ssi).startsWith("//")) {
+            for (index ← ssi + 3 to uri.length) {
+                uri.charAt(index) match {
+                    case '/' | '\\' | '?' | '#' ⇒ return index;
```
Wouldn't this result in a match not exhaustive error if the character is anything else?

errt replied:
Also, isn't there some kind of `indexOf` method that could be used here instead of building this your own?

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -0,0 +1,1467 @@
+        }
+        uri.length
+    }
+    def findPath(
```
Formatting: Missing blank line between methods.

Commit:
Moved AndroidEntryPointsFinder to cg.android.
Refactoring.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -190,17 +184,14 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
-            calls.addCall(contexts(decMeths(e._1._1)), e._1._2, contexts(decMeths(e._2)))
-            resList ++= calls.partialResults(contexts(decMeths(e._1._1))).seq
+            calls.addCall(typeProvider.newContext(decMeths(e._1._1)), e._1._2, typeProvider.newContext(decMeths(e._2)))
+            calls.partialResults(typeProvider.newContext(decMeths(e._1._1))).seq
```
Should probably create this context only once. Which also gives you a chance to introduce a local variable with a descriptive name.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -1328,7 +1279,19 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
+            if (expr.isStringConst) {
+                Some(expr.asStringConst.value)
+            } else if (expr.isGetStatic) {
+                Some(expr.asGetStatic.name)
```
Can you please comment on how it is sensible to use either a string constant or the name of a field (in contrast to, e.g., the field's value)?

Diarilex replied:
You are right the name does not always equal the value. Therfore this will be changed to the fields value. Is the "findField" method of classfile the right way to access it?

errt replied:
No, there just is no proper way to access values in fields. Analyzing this is far from trivial. We do have a points-to analysis that can be used to find some cases with constant values stored in fields, but in general, it is not possible to get all the possible values stored in a field.
So you probably have to see how important this is for your analysis and how you can approximate it.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -1172,11 +1121,13 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
             for (index ← ssi + 3 to uri.length) {
                 uri.charAt(index) match {
                     case '/' | '\\' | '?' | '#' ⇒ return index;
```
Did you check whether there's some `indexOf` or similar method to use instead?

Diarilex replied:
There is an "indexOf" method that can be used instead.

Diarilex committed:
Fixed some mutable datastructures.

Diarilex committed:
Pattern matching.

Diarilex committed:
Merge branch 'develop' into develop

Diarilex committed:
indexOf

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -260,9 +261,8 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
-                for (index ← 1 until lifecycleList.size) {
-                    edges += (lifecycleList(index - 1), 0) -> lifecycleList(index)
-                }
+                edges ++= (for (index ← 1 until lifecycleList.size) yield (lifecycleList(index - 1), 0) -> lifecycleList(index))
```
Please fix formatting, these lines are too long

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -808,12 +806,12 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
         if (intentFilter.registeredReceivers.nonEmpty) {
-            intentFilter.registeredReceivers.map(r ⇒ project.classFile(r)).
-                filter(cf ⇒ cf.isDefined).map(_.get).foreach { addRec ⇒
+            intentFilters ++= intentFilter.registeredReceivers.map(r ⇒ project.classFile(r)).
```
This looks way less efficient than the old variant based on foreach. In general, avoid map and filter operations where the resulting collection will not be used itself. At the very least, use .iterator to avoid the creation intermediate collections.

This applies to several changes in this commit.

errt commented on OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala:
```diff
@@ -1120,28 +1118,22 @@ class AndroidICCAnalysis(val project: SomeProject) extends FPCFAnalysis {
-        uri.substring(start + 1)
+        val i = start + 1
+        val sub = uri.substring(i)
+        val indexList = List('?', '#', '*').map(_.toInt).map(sub.indexOf).filter(_ != -1)
```
This still looks very suspicious. What is the map(_.toInt) doing there? Scala has an indexOf that takes a char (https://www.scala-lang.org/api/2.12.7/scala/collection/immutable/StringOps.html#indexOf(elem:Char):Int)
The combination of map, map, filter and min also looks inefficient. I suggest you use indexWhere to check for any of the three chars at once.
Also, instead of producing the substring sub, you can use the indexOf/indexWhere variants that use a `from` argument (take into account that this changes the semantics of the result though, it is then the index from the beginning of url, not the distance from `i`). You should then also get rid of `i` probably.

TorunR committed:
Merge branch 'develop' into develop
